src/core/collector/mod.rs

```diff
@@ -0,0 +1,757 @@
+// Copyright 2023 Parity Technologies (UK) Ltd.
+// This file is part of polkadot-introspector.
+//
+// polkadot-introspector is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// polkadot-introspector is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with polkadot-introspector.  If not, see <http://www.gnu.org/licenses/>.
+
+use clap::Parser;
+use log::{debug, info, warn};
+use std::{
+	cmp::Ordering,
+	collections::BTreeMap,
+	default::Default,
+	hash::Hash,
+	net::SocketAddr,
+	time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use codec::{Decode, Encode};
+use color_eyre::eyre::eyre;
+use subxt::{
+	ext::{
+		sp_core::{Hasher, H256},
+		sp_runtime::traits::BlakeTwo256,
+	},
+	PolkadotConfig,
+};
+use tokio::sync::{
+	broadcast::{self, Receiver as BroadcastReceiver, Sender as BroadcastSender},
+	mpsc::{error::TryRecvError, Receiver as MspcReceiver},
+};
+
+pub mod candidate_record;
+mod ws;
+
+use crate::core::{
+	ApiService, RecordTime, RecordsStorageConfig, StorageEntry, SubxtCandidateEvent, SubxtCandidateEventType,
+	SubxtDispute, SubxtDisputeResult, SubxtEvent,
+};
+
+use candidate_record::*;
+use ws::*;
+
+#[derive(Clone, Debug, Parser)]
+#[clap(rename_all = "kebab-case")]
+pub struct CollectorOptions {
+	/// Maximum blocks to store
+	#[clap(name = "max-blocks", long)]
+	max_blocks: Option<usize>,
+	/// WS listen address to bind to
+	#[clap(short = 'l', long = "listen")]
+	listen_addr: Option<SocketAddr>,
+}
+
+/// This type is used to distinguish different keys in the storage
+#[derive(Clone, Copy, Debug, Hash, Ord, PartialOrd, Eq, PartialEq)]
+pub enum CollectorPrefixType {
+	/// Candidate prefixed by Parachain-Id
+	Candidate(u32),
+	/// A mapping to find out parachain id by candidate hash (e.g. when we don't know the parachain)
+	CandidatesParachains,
+	/// Relay chain block
+	Head,
+	/// Session data, keyed by session hash (blake2b(session_index))
+	Session,
+	/// Inherent data (more expensive to store, so good to have it shared)
```

Comment:
I think we have to be careful here and adjust the prune settings. Right now we keep stuff for 1000 blocks, so we end up with high memory requirements for running in Prometheus mode without actually needing it.

Reply:
It is now 64 blocks by default, so it shouldn't be a big issue (under 100M of memory I suppose).